cudapoa/tests/Test_CudapoaNW.cu

```diff
@@ -304,6 +300,213 @@ TEST_P(NWTest, TestNWCorrectness)
+                      alignment_read,
+                      batch_size.alignment_band_width,
+                      batch_size.max_banded_pred_distance,
+                      gap_score,
+                      mismatch_score,
+                      match_score,
+                      aligned_nodes,
+                      adaptive);
+    }
+    else
+    {
+        runNWbanded(nodes,
+                    graph,
+                    node_id_to_pos,
+                    graph_count,
+                    incoming_edge_count,
+                    incoming_edges,
+                    outgoing_edge_count,
+                    read,
+                    read_count,
+                    scores,
+                    batch_size.matrix_sequence_dimension,
+                    batch_size.max_nodes_per_graph,
+                    alignment_graph,
+                    alignment_read,
+                    batch_size.alignment_band_width,
+                    gap_score,
+                    mismatch_score,
+                    match_score,
+                    aligned_nodes,
+                    adaptive);
+    }
+
+    GW_CU_CHECK_ERR(cudaDeviceSynchronize());
+
+    //input and output buffers are the same ones in unified memory, so the results are updated in place
+    //results are stored in alignment_graph and alignment_read; return string representation of those
+    auto res = std::make_pair(genomeworks::stringutils::array_to_string<int16_t>(alignment_graph, *aligned_nodes, ","),
+                              genomeworks::stringutils::array_to_string<int16_t>(alignment_read, *aligned_nodes, ","));
+
+    GW_CU_CHECK_ERR(cudaFree(nodes));
+    GW_CU_CHECK_ERR(cudaFree(graph));
+    GW_CU_CHECK_ERR(cudaFree(node_id_to_pos));
+    GW_CU_CHECK_ERR(cudaFree(incoming_edges));
+    GW_CU_CHECK_ERR(cudaFree(incoming_edge_count));
+    GW_CU_CHECK_ERR(cudaFree(outgoing_edges));
+    GW_CU_CHECK_ERR(cudaFree(outgoing_edge_count));
+    GW_CU_CHECK_ERR(cudaFree(scores));
+    GW_CU_CHECK_ERR(cudaFree(alignment_graph));
+    GW_CU_CHECK_ERR(cudaFree(read));
+    GW_CU_CHECK_ERR(cudaFree(alignment_read));
+    GW_CU_CHECK_ERR(cudaFree(aligned_nodes));
+    if (traceback)
+    {
+        GW_CU_CHECK_ERR(cudaFree(traces));
+    }
+
+    return res;
+}
+
+class NWbandedTest : public ::testing::Test
+{
+public:
+    std::unique_ptr<BasicNW> nw;
+
+public:
+    void SetUp()
+    {
+        // initialize nw graph and read with the following data
+        std::string nodes_str = "TTTAACCTAATAAATCAGTGAAGATTTAAAATATGATAATTATTGATTTTGGTGAGAGTGCAAAGAAATTTGTTACCCTCATAAGCTGAGCAGACAGATAAGATAGAAAAACAGAAGATAGAATATTAAAACCATGATAGGTACAGACTGAAAAATTCTTGGATAAATATTAAAATTTAGGCTTTAGTAGTAGATTGATGACTGTGAGGAAAAAGGATGTCCAATTGTTGAGTGACATGTAGAATGCCTTAAAATAATTTTACACGTCACTGAAAGCTATATTTATATTCAGGAAGGATATATCCCAGTCATGATTTTCTTAATAAGTTGCCCCATTTTCCAAGTTTAGCTAATTAACATTTATGTCTTCTATAATCAGGAATAGTCATTAACTGACACAGAAACAATTGGAAGCATATGTAGCCAAAAACATAAAAATTATTGCATCCAAATAATGATAAAGTAAAATATTAAAAAATATAGTCTTCTAAAT";
```

Comment:
Optional: I currently see this as a single test case for this suite. Will it be possible to modularize it out so that we can add additional test cases for corner cases we see later?

Reply:
This is implemented as `TEST_F`, test with fixtures, which allow multiple tests use the same initialization which takes place in `SetUp()`. This was suitable for this case where we have 4 tests for validating 4 different band modes against full-alignment. All 4 using the same setup.

Reply:
I understand. I was recommending parameterizing the two strings used to create the graph.

Reply:
This is an extension to the previous NW test which used only a very short string (4-5 bases). In the even of designing new tests for NW that could use this setup, they can be added as new `TEST_F` cases, otherwise they can use e.g. `sample_windows.txt`. Therefore I'd rather leave it unchanged.